source/dofs/dof_tools_constraints.cc

```diff
@@ -1070,6 +1101,1316 @@ namespace DoFTools
+              const unsigned int n_external_lines_on_children = 8;
+
+              const unsigned int n_lines_on_children =
+                n_intern_lines_on_children + n_external_lines_on_children;
+
+              // we only consider the isotropic case here
+              const unsigned int n_of_children = 4;
+
+              // number of DoFs on the children
+              // Remark: Nedelec elements have no DoFs on the vertices,
+              // therefore we skip the vertices
+              const unsigned int n_dofs_on_children =
+                (n_lines_on_children * fe.n_dofs_per_line() +
+                 n_of_children * fe.n_dofs_per_quad(face));
+
+              // we might not use all of those in case of artificial cells, so
+              // do not resize(), but reserve() and use push_back later.
+              dofs_on_children.clear();
+              dofs_on_children.reserve(n_dofs_on_children);
+
+              Assert(n_dofs_on_mother == fe.constraints().n(),
+                     ExcDimensionMismatch(n_dofs_on_mother,
+                                          fe.constraints().n()));
+              Assert(n_dofs_on_children == fe.constraints().m(),
+                     ExcDimensionMismatch(n_dofs_on_children,
+                                          fe.constraints().m()));
+
+              // get the current face
+              const typename DoFHandler<3, 3>::face_iterator this_face =
+                cell->face(face);
+
+              // fill the DoFs on the mother:
+              unsigned int next_index = 0;
+
+              // DoFs on vertices:
+              // Nedelec elements have no DoFs on the vertices
+
+              // DoFs on lines:
+              for (unsigned int line = 0;
+                   line < GeometryInfo<dim>::lines_per_face;
+                   ++line)
+                for (unsigned int dof = 0; dof != fe.n_dofs_per_line(); ++dof)
+                  dofs_on_mother[next_index++] =
+                    this_face->line(line)->dof_index(dof, fe_index);
+
+              // DoFs on the face:
+              for (unsigned int dof = 0; dof != fe.n_dofs_per_quad(face); ++dof)
+                dofs_on_mother[next_index++] =
+                  this_face->dof_index(dof, fe_index);
+
+              // check that we have added all DoFs
+              AssertDimension(next_index, dofs_on_mother.size());
+
+              // fill the DoF on the children:
+              Assert(dof_handler.get_triangulation()
+                         .get_anisotropic_refinement_flag() ||
+                       ((this_face->child(0)->vertex_index(3) ==
+                         this_face->child(1)->vertex_index(2)) &&
+                        (this_face->child(0)->vertex_index(3) ==
+                         this_face->child(2)->vertex_index(1)) &&
+                        (this_face->child(0)->vertex_index(3) ==
+                         this_face->child(3)->vertex_index(0))),
+                     ExcInternalError());
```

Comment:
What's going on here? Why is it OK if there has been some anisotropic refinement?

Reply:
Well, this is a copy-and-paste mistake. I added more asserts directly before opening this pull request. I intend to check that we have only isotropic refinement, but I forgot to adapt that assertion accordingly.
I have fixed that.